various fixes to correct things preventing argo from deploying properly to stage

jmartin-sul opened the pull request:
(similar changes will soon be needed for all other instances).

* _deploy.rb_: `config/environments/` in `:linked_dirs` becomes `config/settings/`, since this is now where per-instance config info lives.
* _deploy.rb_: the post-restart trigger for running the `argo:htaccess` rake task has been removed, as web code deployments don't necessarily coincide with the need for .htaccess refreshing, and the task is run regularly by a cron job anyway.
* _dor_staging.rb_: full paths to cert file and key file are now specified, since they no longer necessarily share a parent directory.  NOTE: a similar change will need to be made in dor_development.rb and dor_production.rb.
* _argo.rake_: `argo:htaccess` rake task now always includes the built-in argo workgroups, and de-dupes the workgroup list.
* _argo.rake_: `argo:htaccess` rake task now has some usage explanation.

darrenleeweber commented:
Maybe add documentation in the deploy files on where, how and why the htaccess directives are managed in puppet?

darrenleeweber commented on config/deploy.rb:
```diff
@@ -25,10 +25,10 @@
 
 # Default value for :linked_files is []
-set :linked_files, %w{config/database.yml config/solr.yml config/default_htaccess_directives}
+set :linked_files, %w{config/database.yml config/solr.yml config/settings/staging.yml}
```
Not sure that `config/deploy.rb` should be tied to `settings/staging.yml` in particular.  Maybe that belongs in a subdirectory or another place that is specific to staging?

mejackreed replied:
Good catch, agree

jmartin-sul replied:
yup, good catch, will fix.

darrenleeweber commented on lib/tasks/argo.rake:
```diff
@@ -279,29 +279,6 @@ namespace :argo do
-      File.open(File.join(Rails.root, 'public/.htaccess'),'w') do |htaccess|
-        htaccess.puts directives.sort.join("\n")
-      end
-      File.unlink('public/auth/.htaccess') if File.exist?('public/auth/.htaccess')
```
Can puppet replace all this, esp. whatever `get_workgroups_facet` is doing?

jmartin-sul replied:
@eefahy pointed out that it's unnecessarily circuitous.  this code queries the existing objects for permissions related info, and builds up a list from the permissions that are currently in use.  but that's unnecessary, since we know all the workgroups we'll care about supporting.  and now that i think about it, we might want to add them to the htaccess file before they're listed in an APO.  if we define new workgroups, which we do occasionally, we can manually add them to whatever config puppet uses.  and as per the commit message, updating htaccess doesn't really correspond with deploying/restarting the app, which is when this rake task is run.

it's clever, but it doesn't really gain us anything, as far as we can tell.

mejackreed commented on config/deploy.rb:
```diff
@@ -25,10 +25,10 @@
 
 # Default value for linked_dirs is []
-set :linked_dirs, %w{log config/certs config/environments tmp/pids tmp/cache tmp/sockets vendor/bundle public/system}
+set :linked_dirs, %w{log config/certs tmp/pids tmp/cache tmp/sockets vendor/bundle public/system}
```
We should probably have `config/settings` here. See: https://github.com/sul-dlss/earthworks/blob/master/config/deploy.rb#L30

jmartin-sul replied:
good point.  could then get rid of `config/settings/staging.yml` from `:linked_files`.

atz commented:
Probably should squash part or all of this?

darrenleeweber commented:
Some mysteries solved, but maybe logging the workgroups would help?  I'm dying to know more about `my-peeps`!

jmartin-sul commented:
@atz yup, this is work in progress.  will squash when done.

@darrenleeweber i do like the idea of logging, though that might get a little noisy if the task is running every few minutes.  also, it'd mostly be useful for historical comparison purposes, since the htaccess output is essentially the same as what would be logged (though i can see how the history could be useful, and it's not like the list is _that_ long).

mejackreed commented:
Should we also set to `set :bundle_without, %w{test development}` https://github.com/sul-dlss/argo/blob/develop/config/deploy/stage.rb#L5

rather than production.

jmartin-sul commented:
@atz: squashed (and rebased from develop).  diff looks good, and github says it can merge it automatically, so i think it's ok (i pulled when i shouldn't have after squashing, but i think i got myself out of that jam by resetting HEAD to my squash commit before force pushing).

@mejackreed: fixed `:bundle_without`.

@darrenleeweber: added explanation for how the .htaccess rake task is invoked.  declined to log the list used for htaccess generation.  if you feel strongly about keeping a historical record, i'd be more inclined to create a new issue/PR, and keep N backups of the .htaccess file, maybe capturing only when the new file actually differs from the old one.  but i think i'd want to discuss the how/why of doing that, and it feels outside the scope of this PR.

i think that addresses all the issues raised so far with this one?
